### PR TITLE
Update Node.js to v18 and npm to v9

### PR DESCRIPTION
**Pull request by lucferbux**
## Description

Closes: #915 

## How Has This Been Tested?

1. Run the installation locally
2. Create a new build
3. Test that the build is working

Image: `quay.io/lferrnan/odh-dashboard:rhods-915`

**You should bump your local development to node 18 (nvm use 18) and npm v9**
**In some linux distributions npm +7 routes automatically with ipv6, so you might need to change your network config**

## Merge criteria:

- [X] The commits are squashed in a cohesive manner and have meaningful messages.
- [X] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [X] The developer has manually tested the changes and verified that the changes work

### CODE REVIEW

**Comment by lucferbux** [comment]
There are still a lot of pair dependencies from `@openshift/dynamic-plugin-sdk` that are conflicting our packages, as this could be a long process, I've modified the script to force installation in the frontend. We are having this same conflicts in our current setup but npm hides them, I think we should go on and fix these dependency errors in another PR.

What do you think @andrewballantyne @DaoDaoNoCode 

```bash
npm WARN using --force Recommended protections disabled.
npm WARN ERESOLVE overriding peer dependency
npm WARN While resolving: @openshift/dynamic-plugin-sdk-extensions@1.2.0
npm WARN Found: react-redux@8.0.5
npm WARN node_modules/react-redux
npm WARN   react-redux@"^8.0.4" from the root project
npm WARN 
npm WARN Could not resolve dependency:
npm WARN peerOptional react-redux@"^7.2.2" from @openshift/dynamic-plugin-sdk-extensions@1.2.0
npm WARN node_modules/@openshift/dynamic-plugin-sdk-extensions
npm WARN   peer @openshift/dynamic-plugin-sdk-extensions@"^1.2.0" from @openshift/dynamic-plugin-sdk-utils@1.2.0
npm WARN   node_modules/@openshift/dynamic-plugin-sdk-utils
npm WARN 
npm WARN Conflicting peer dependency: react-redux@7.2.9
npm WARN node_modules/react-redux
npm WARN   peerOptional react-redux@"^7.2.2" from @openshift/dynamic-plugin-sdk-extensions@1.2.0
npm WARN   node_modules/@openshift/dynamic-plugin-sdk-extensions
npm WARN     peer @openshift/dynamic-plugin-sdk-extensions@"^1.2.0" from @openshift/dynamic-plugin-sdk-utils@1.2.0
npm WARN     node_modules/@openshift/dynamic-plugin-sdk-utils
npm WARN ERESOLVE overriding peer dependency
npm WARN While resolving: @openshift/dynamic-plugin-sdk-extensions@1.2.0
npm WARN Found: react-router@6.8.2
npm WARN node_modules/react-router
npm WARN   react-router@"^6.4.1" from the root project
npm WARN   1 more (react-router-dom)
npm WARN 
npm WARN Could not resolve dependency:
npm WARN peerOptional react-router@"^5.2.1" from @openshift/dynamic-plugin-sdk-extensions@1.2.0
npm WARN node_modules/@openshift/dynamic-plugin-sdk-extensions
npm WARN   peer @openshift/dynamic-plugin-sdk-extensions@"^1.2.0" from @openshift/dynamic-plugin-sdk-utils@1.2.0
npm WARN   node_modules/@openshift/dynamic-plugin-sdk-utils
npm WARN 
npm WARN Conflicting peer dependency: react-router@5.3.4
npm WARN node_modules/react-router
npm WARN   peerOptional react-router@"^5.2.1" from @openshift/dynamic-plugin-sdk-extensions@1.2.0
npm WARN   node_modules/@openshift/dynamic-plugin-sdk-extensions
npm WARN     peer @openshift/dynamic-plugin-sdk-extensions@"^1.2.0" from @openshift/dynamic-plugin-sdk-utils@1.2.0
npm WARN     node_modules/@openshift/dynamic-plugin-sdk-utils
npm WARN ERESOLVE overriding peer dependency
npm WARN While resolving: @openshift/dynamic-plugin-sdk-utils@1.2.0
npm WARN Found: react-redux@8.0.5
npm WARN node_modules/react-redux
npm WARN   react-redux@"^8.0.4" from the root project
npm WARN 
npm WARN Could not resolve dependency:
npm WARN peer react-redux@"^7.2.2" from @openshift/dynamic-plugin-sdk-utils@1.2.0
npm WARN node_modules/@openshift/dynamic-plugin-sdk-utils
npm WARN   @openshift/dynamic-plugin-sdk-utils@"^1.2.0" from the root project
npm WARN 
npm WARN Conflicting peer dependency: react-redux@7.2.9
npm WARN node_modules/react-redux
npm WARN   peer react-redux@"^7.2.2" from @openshift/dynamic-plugin-sdk-utils@1.2.0
npm WARN   node_modules/@openshift/dynamic-plugin-sdk-utils
npm WARN     @openshift/dynamic-plugin-sdk-utils@"^1.2.0" from the root project
npm WARN ERESOLVE overriding peer dependency
npm WARN While resolving: @openshift/dynamic-plugin-sdk-utils@1.2.0
npm WARN Found: react-router@6.8.2
npm WARN node_modules/react-router
npm WARN   react-router@"^6.4.1" from the root project
npm WARN   1 more (react-router-dom)
npm WARN 
npm WARN Could not resolve dependency:
npm WARN peer react-router@"^5.2.1" from @openshift/dynamic-plugin-sdk-utils@1.2.0
npm WARN node_modules/@openshift/dynamic-plugin-sdk-utils
npm WARN   @openshift/dynamic-plugin-sdk-utils@"^1.2.0" from the root project
npm WARN 
npm WARN Conflicting peer dependency: react-router@5.3.4
npm WARN node_modules/react-router
npm WARN   peer react-router@"^5.2.1" from @openshift/dynamic-plugin-sdk-utils@1.2.0
npm WARN   node_modules/@openshift/dynamic-plugin-sdk-utils
npm WARN     @openshift/dynamic-plugin-sdk-utils@"^1.2.0" from the root project
npm WARN ERESOLVE overriding peer dependency
npm WARN ERESOLVE overriding peer dependency
npm WARN ERESOLVE overriding peer dependency
npm WARN ERESOLVE overriding peer dependency
npm WARN ERESOLVE overriding peer dependency
npm WARN While resolving: msw-storybook-addon@1.7.0
npm WARN Found: msw@1.1.0
npm WARN node_modules/msw
npm WARN   dev msw@"^1.1.0" from the root project
npm WARN 
npm WARN Could not resolve dependency:
npm WARN peer msw@">=0.35.0 <1.0.0" from msw-storybook-addon@1.7.0
npm WARN node_modules/msw-storybook-addon
npm WARN   dev msw-storybook-addon@"^1.7.0" from the root project
npm WARN 
npm WARN Conflicting peer dependency: msw@0.49.3
npm WARN node_modules/msw
npm WARN   peer msw@">=0.35.0 <1.0.0" from msw-storybook-addon@1.7.0
npm WARN   node_modules/msw-storybook-addon
npm WARN     dev msw-storybook-addon@"^1.7.0" from the root project
npm WARN ERESOLVE overriding peer dependency
npm WARN While resolving: react-virtualized@9.22.3
npm WARN Found: react@18.2.0
npm WARN node_modules/react
npm WARN   peer react@"^16.8.4 || ^17.0.0" from react-inspector@5.1.1
npm WARN   node_modules/@storybook/addon-actions/node_modules/react-inspector
npm WARN     react-inspector@"^5.1.0" from @storybook/addon-actions@6.5.16
npm WARN     node_modules/@storybook/addon-actions
npm WARN 
npm WARN Could not resolve dependency:
npm WARN peer react@"^15.3.0 || ^16.0.0-alpha" from react-virtualized@9.22.3
npm WARN node_modules/react-virtualized
npm WARN   peer react-virtualized@"^9.22.3" from @openshift/dynamic-plugin-sdk-utils@1.2.0
npm WARN   node_modules/@openshift/dynamic-plugin-sdk-utils
npm WARN   1 more (@patternfly/react-virtualized-extension)
npm WARN 
npm WARN Conflicting peer dependency: react@16.14.0
npm WARN node_modules/react
npm WARN   peer react@"^15.3.0 || ^16.0.0-alpha" from react-virtualized@9.22.3
npm WARN   node_modules/react-virtualized
npm WARN     peer react-virtualized@"^9.22.3" from @openshift/dynamic-plugin-sdk-utils@1.2.0
npm WARN     node_modules/@openshift/dynamic-plugin-sdk-utils
npm WARN     1 more (@patternfly/react-virtualized-extension)
npm WARN ERESOLVE overriding peer dependency
npm WARN While resolving: react-virtualized@9.22.3
npm WARN Found: react-dom@18.2.0
npm WARN node_modules/react-dom
npm WARN   peer react-dom@"^0.14.8 || ^15.0.1 || ^16.0.0 || ^17.0.1" from react-element-to-jsx-string@14.3.4
npm WARN   node_modules/@storybook/react/node_modules/react-element-to-jsx-string
npm WARN     react-element-to-jsx-string@"^14.3.4" from @storybook/react@6.5.16
npm WARN     node_modules/@storybook/react
npm WARN 
npm WARN Could not resolve dependency:
npm WARN peer react-dom@"^15.3.0 || ^16.0.0-alpha" from react-virtualized@9.22.3
npm WARN node_modules/react-virtualized
npm WARN   peer react-virtualized@"^9.22.3" from @openshift/dynamic-plugin-sdk-utils@1.2.0
npm WARN   node_modules/@openshift/dynamic-plugin-sdk-utils
npm WARN   1 more (@patternfly/react-virtualized-extension)
npm WARN 
npm WARN Conflicting peer dependency: react-dom@16.14.0
npm WARN node_modules/react-dom
npm WARN   peer react-dom@"^15.3.0 || ^16.0.0-alpha" from react-virtualized@9.22.3
npm WARN   node_modules/react-virtualized
npm WARN     peer react-virtualized@"^9.22.3" from @openshift/dynamic-plugin-sdk-utils@1.2.0
```

**Review by DaoDaoNoCode — COMMENTED** [review]
I tried to run it locally, the `npm install` succeeded, however, when I run the server, the backend crashed with
```
[dev:backend ] Error: Cannot find module 'node:fs'
[dev:backend ] Require stack:
[dev:backend ] - /Users/juntaowang/Desktop/odh-dashboard/backend/node_modules/@fastify/send/lib/SendStream.js
[dev:backend ] - /Users/juntaowang/Desktop/odh-dashboard/backend/node_modules/@fastify/send/index.js
[dev:backend ] - /Users/juntaowang/Desktop/odh-dashboard/backend/node_modules/@fastify/static/index.js
[dev:backend ] - /Users/juntaowang/Desktop/odh-dashboard/backend/src/app.ts
[dev:backend ] - /Users/juntaowang/Desktop/odh-dashboard/backend/src/server.ts
```
And the frontend has an error too:
```
[dev:frontend] <i> [webpack-dev-middleware] wait until bundle finished: /
[dev:frontend] ERROR in ./node_modules/@openshift/dynamic-plugin-sdk-utils/dist/index.js 24:0-75
[dev:frontend] Module not found: Error: Can't resolve '@openshift/dynamic-plugin-sdk-extensions' in '/Users/juntaowang/Desktop/odh-dashboard/frontend/node_modules/@openshift/dynamic-plugin-sdk-utils/dist'
[dev:frontend]  @ ./src/SDKInitialize.tsx 3:0-83 61:16-32 71:24-34
[dev:frontend]  @ ./src/index.tsx 7:0-44 20:32-45
[dev:frontend] 
[dev:frontend] webpack compiled with 1 error
```

**Comment by lucferbux** [comment]
> I tried to run it locally, the `npm install` succeeded, however, when I run the server, the backend crashed with
> 
> ```
> [dev:backend ] Error: Cannot find module 'node:fs'
> [dev:backend ] Require stack:
> [dev:backend ] - /Users/juntaowang/Desktop/odh-dashboard/backend/node_modules/@fastify/send/lib/SendStream.js
> [dev:backend ] - /Users/juntaowang/Desktop/odh-dashboard/backend/node_modules/@fastify/send/index.js
> [dev:backend ] - /Users/juntaowang/Desktop/odh-dashboard/backend/node_modules/@fastify/static/index.js
> [dev:backend ] - /Users/juntaowang/Desktop/odh-dashboard/backend/src/app.ts
> [dev:backend ] - /Users/juntaowang/Desktop/odh-dashboard/backend/src/server.ts
> ```
> 
> And the frontend has an error too:
> 
> ```
> [dev:frontend] <i> [webpack-dev-middleware] wait until bundle finished: /
> [dev:frontend] ERROR in ./node_modules/@openshift/dynamic-plugin-sdk-utils/dist/index.js 24:0-75
> [dev:frontend] Module not found: Error: Can't resolve '@openshift/dynamic-plugin-sdk-extensions' in '/Users/juntaowang/Desktop/odh-dashboard/frontend/node_modules/@openshift/dynamic-plugin-sdk-utils/dist'
> [dev:frontend]  @ ./src/SDKInitialize.tsx 3:0-83 61:16-32 71:24-34
> [dev:frontend]  @ ./src/index.tsx 7:0-44 20:32-45
> [dev:frontend] 
> [dev:frontend] webpack compiled with 1 error
> ```

Have you tried to manually delete both `node_modules` folders?


And please let me know if you are using `npm 9.5.0` or higher.

Just tested it on a fresh install in a new machine and cleaning all in my local machine and it's working as expected.

**Review by Gkrumbach07 — APPROVED** [review]
installs, runs, builds properly

/lgtm

**Comment by DaoDaoNoCode** [comment]
@lucferbux Hi, I am still seeing this error in my backend:
```
[dev:backend ] Error: Cannot find module 'node:fs'
[dev:backend ] Require stack:
[dev:backend ] - /Users/juntaowang/Desktop/odh-dashboard/backend/node_modules/@fastify/send/lib/SendStream.js
[dev:backend ] - /Users/juntaowang/Desktop/odh-dashboard/backend/node_modules/@fastify/send/index.js
[dev:backend ] - /Users/juntaowang/Desktop/odh-dashboard/backend/node_modules/@fastify/static/index.js
[dev:backend ] - /Users/juntaowang/Desktop/odh-dashboard/backend/src/app.ts
[dev:backend ] - /Users/juntaowang/Desktop/odh-dashboard/backend/src/server.ts
[dev:backend ]     at Function.Module._resolveFilename (internal/modules/cjs/loader.js:885:15)
[dev:backend ]     at Function.Module._resolveFilename.sharedData.moduleResolveFilenameHook.installedValue [as _resolveFilename] (/Users/juntaowang/Desktop/odh-dashboard/backend/node_modules/@cspotcode/source-map-support/source-map-support.js:811:30)
[dev:backend ]     at Function.Module._load (internal/modules/cjs/loader.js:730:27)
[dev:backend ]     at Module.require (internal/modules/cjs/loader.js:957:19)
[dev:backend ]     at require (internal/modules/cjs/helpers.js:88:18)
[dev:backend ]     at Object.<anonymous> (/Users/juntaowang/Desktop/odh-dashboard/backend/node_modules/@fastify/send/lib/SendStream.js:10:12)
[dev:backend ]     at Module._compile (internal/modules/cjs/loader.js:1068:30)
[dev:backend ]     at Module._extensions..js (internal/modules/cjs/loader.js:1097:10)
[dev:backend ]     at Object.require.extensions.<computed> [as .js] (/Users/juntaowang/Desktop/odh-dashboard/backend/node_modules/ts-node/src/index.ts:1608:43)
[dev:backend ]     at Module.load (internal/modules/cjs/loader.js:933:32) {
[dev:backend ]   code: 'MODULE_NOT_FOUND',
[dev:backend ]   requireStack: [
[dev:backend ]     '/Users/juntaowang/Desktop/odh-dashboard/backend/node_modules/@fastify/send/lib/SendStream.js',
[dev:backend ]     '/Users/juntaowang/Desktop/odh-dashboard/backend/node_modules/@fastify/send/index.js',
[dev:backend ]     '/Users/juntaowang/Desktop/odh-dashboard/backend/node_modules/@fastify/static/index.js',
[dev:backend ]     '/Users/juntaowang/Desktop/odh-dashboard/backend/src/app.ts',
[dev:backend ]     '/Users/juntaowang/Desktop/odh-dashboard/backend/src/server.ts'
[dev:backend ]   ]
[dev:backend ] }
[dev:backend ] [nodemon] app crashed - waiting for file changes before starting...
```
Looks like it happens in the fastify packages. I already deleted all the node_modules and re-install them again. My current npm version is `9.4.0` and the node version is `18.15.0`.

**Review by DaoDaoNoCode — APPROVED** [review]
After I deleted all the `package-lock.json` files, the `npm run dev` works! Everything works as expected, lgtm

**Comment by andrewballantyne** [comment]
> After I deleted all the `package-lock.json` files, the `npm run dev` works! Everything works as expected, lgtm

Deleting lock files is never a solution we should take lightly. They are there for a reason -- they make sure every dependency in all aspects are sourced exactly the same for each one of us.

**Review by DaoDaoNoCode — APPROVED** [review]
Tested with the latest change, it works well on the dev side.

**Comment by lucferbux** [comment]
Tested in `arm64` & `x64` macos and fedora. Building Image can be found here: `quay.io/lferrnan/odh-dashboard:rhods-915`.

File: frontend/package.json [hunk 1]
```diff
@@ -117,7 +117,7 @@
     "jest-environment-jsdom": "^29.4.3",
     "js-yaml-loader": "^1.2.2",
     "mini-css-extract-plugin": "^2.7.2",
-    "msw": "^1.1.0",
+    "msw": "^0.49.3",
```
Comment (andrewballantyne) [inline frontend/package.json:120]: woah -- down versioning?

File: frontend/package.json [hunk 2]
```diff
@@ -38,9 +38,9 @@
     "storybook": "start-storybook -p 6006 -s ./config"
   },
   "dependencies": {
-    "@openshift/dynamic-plugin-sdk": "^1.0.0-alpha12",
-    "@openshift/dynamic-plugin-sdk-utils": "^1.0.0-alpha17",
     "@patternfly/patternfly": "4.224.2",
+    "@openshift/dynamic-plugin-sdk": "^3.0.0",
+    "@openshift/dynamic-plugin-sdk-utils": "^1.2.0",
```
Comment (andrewballantyne) [inline frontend/package.json:43]: Why did you move these -- this should be in alphabetical order? Did you manually do this because I would assume NPM sorts them...

Comment (lucferbux) [reply frontend/package.json:43]: Yes, my bad, sorting it again.

**Review by andrewballantyne — APPROVED** [review]
Here we go! Let's make sure we get this into RHODS asap as well to make sure no build image issues crop up because of this.
